File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def find_longest(self, i: int, j: int, n: int, m: int, text1: str, text2: str, dp: list) -> int:
        if i >= n or j >= m:
            return 0  # Base case: end of either string
        
        if dp[i][j] != -1:
            return dp[i][j]  # Return cached result if already computed

        match = 0
        not_match1 = 0
        not_match2 = 0
        
        if text1[i] != text2[j]:
            # Explore both possibilities: skip a character from either string
            not_match1 = self.find_longest(i + 1, j, n, m, text1, text2, dp)
            not_match2 = self.find_longest(i, j + 1, n, m, text1, text2, dp)
        else:
            # If the characters match, consider it as part of LCS and move to the next characters in both strings
            match = 1 + self.find_longest(i + 1, j + 1, n, m, text1, text2, dp)
        
        dp[i][j] = max(match, not_match1, not_match2)  # Memoize the result
        return dp[i][j]
    
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n = len(text1)
        m = len(text2)
        dp = [[-1] * m for _ in range(n)]  # Initialize memoization table with -1
        return self.find_longest(0, 0, n, m, text1, text2, dp)

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
private:
    int lengthOfLCS(int ind1, int ind2, int n, int m, string& text1, string& text2,vector<vector<int>>& dp)
    {
        if(ind1 == n || ind2 == m) return 0;
        if(dp[ind1][ind2]!=-1) return dp[ind1][ind2];
        if(text1[ind1] == text2[ind2])
        {
            return dp[ind1][ind2] = 1 + lengthOfLCS(ind1+1,ind2+1,n,m,text1,text2,dp);
        }
        else
        {
            return dp[ind1][ind2] = max(lengthOfLCS(ind1+1,ind2,n,m,text1,text2,dp),lengthOfLCS(ind1, ind2+1,n,m,text1,text2,dp));
        }
        return 0;

    }
public:
    int longestCommonSubsequence(string text1, string text2) {
        int n = text1.size();
        int m = text2.size();
        vector<vector<int>> dp(n+1,vector<int>(m+1,-1));
        return lengthOfLCS(0,0,n,m,text1,text2,dp);
    }
};